sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -202,3 +202,25 @@ object Unions {
     }
   }
 }
+
+/**
+ * A pattern that finds the original expression from a sequence of casts.
+ */
+object Casts {
+  def unapply(expr: Expression): Option[Attribute] = expr match {
+    case c: Cast => collectCasts(expr)
+    case _ => None
+  }
+
+  private def collectCasts(e: Expression): Option[Attribute] = {
```

Comment:
Can use pattern matching to simplify this block. Perhaps something along the lines of:

```scala
  private def collectCasts(expr: Expression): Option[Attribute] = {
    expr match {
      case e: Cast => collectCasts(e.child)
      case e: Attribute => Some(e)
      case _ => None
    }
  }
```